### ISSUE TITLE
How to recover all the links (even manually) in containers deleted by mozilla bug ?

### ISSUE THREAD

**Issue by FreeWF** [mozilla/multi-account-containers#1409]
I lost more than 300 tabs which I recently transfered into custom containers. But after mozilla's new version, all custom containers have been deleted. Mozilla explains their new version won't recover those. But I absolutely need to recover the tabs / links inside. Don't need the containers to be back exactly as is, but I need to recover the links that were inside those custom containers. Can you tell me in which file their are logged, or any manual procedure to have access to my saved tabs's links ? Thanks.

**Comment by calegario** [comment]
> 
> 
> I lost more than 300 tabs which I recently transfered into custom containers. But after mozilla's new version, all custom containers have been deleted. Mozilla explains their new version won't recover those. But I absolutely need to recover the tabs / links inside. Don't need the containers to be back exactly as is, but I need to recover the links that were inside those custom containers. Can you tell me in which file their are logged, or any manual procedure to have access to my saved tabs's links ? Thanks.

If you look at the profile folder (usually C:\Users\@user\AppData\Roaming\Mozilla\Firefox\Profiles), you will find a folder called "browser-extention-data" and finally "@testpilot-containers". The "storage.js.migrated" file contains all the URL's associated to the previous containers.

If you recreate new containers with the same name, all the URL's will be associated to the new ones.

**Comment by FreeWF** [comment]
Than you very much, this is very interesting. I found the file, it contains 692 Kb, but when opening it, URLs are not readable directly, It's probably compressed and / or coded.  Or a special software is needed ? 
I used Sublime text because it is said that this free software can open hex files, but I'm not a programmer at all, so I don't know how to use it properly and how to get the URL in text mode.

> If you recreate new containers with the same name, all the URL's will be associated to the new ones.

This is great news, This could resolve everything if it worked...But how to remember the exact name of each of the 20 containers I created ? Unfortunately mine are not "only one word" type. Maybe the name of the old containers can be also found in a file ? But which file ?

Thanks a lot very, very much in advance for you help. This could help a lot of people for sure.

**Comment by ALTanir** [comment]
Keep back up your OS-C'Drive>AppData> Local&Roaming "Mozilla"-Folder. After test approval and signed properly of the add-ons we can re-enable it. Yes, some add-ons disabled caused corporal certificate expired. And so its need the other versions from list or, email authors to test and sign the versions. Some extensions have already resigned and can recovered. For me -test pilot-4.1.0@fx56.0b5.

**Comment by FreeWF** [comment]
OK, I'll back up the files. Isn't it possible to find the name of my old containers in a specific file ? This may solve the problem, as explained calegario above.  I would have liked to try...

**Comment by calegario** [comment]
> 
> 
> Than you very much, this is very interesting. I found the file, it contains 692 Kb, but when opening it, URLs are not readable directly, It's probably compressed and / or coded. Or a special software is needed ?
> I used Sublime text because it is said that this free software can open hex files, but I'm not a programmer at all, so I don't know how to use it properly and how to get the URL in text mode.
> 
> > If you recreate new containers with the same name, all the URL's will be associated to the new ones.
> 
> This is great news, This could resolve everything if it worked...But how to remember the exact name of each of the 20 containers I created ? Unfortunately mine are not "only one word" type. Maybe the name of the old containers can be also found in a file ? But which file ?
> 
> Thanks a lot very, very much in advance for you help. This could help a lot of people for sure.

You're welcome! I imagine the worry due to this bug.

So, in my case. I used the notepad to open this file like a .txt. It works, but my file is smaller. I recommend Notepad++ as an alternative. You need only a txt reader.

**Comment by calegario** [comment]
> 
> 
> OK, I'll back up the files. Isn't it possible to find the name of my old containers in a specific file ? This may solve the problem, as explained calegario above. I would have liked to try...

On the profile folder (C:\Users@user\AppData\Roaming\Mozilla\Firefox\Profiles\@profile) you will find a file called "containers.json". It's important because it contains the containers names and ID's - that you can associate it to the URL's in the first file mentioned (storage.js.migrated). If you're lucky, you'll find the names and ID's. If it's not possible you can manually associate new containers ID's to the URL's in "storage.js.migrated". 

To summarize: each container had an ID in "container.json". Each URL is associated to this ID in "storage.js.migrated". 

These two files are the most important ones.

**Comment by FreeWF** [comment]
Thanks again Calegario ! My"storage.js.migrated" when opened with notepad isn't directly readable as text. It is probably compressed, they are chunks that you can read, but it's mainly unsuable as is. About "containers.json", it is 1 Kb, not empty but it doesn't contain my custom containers's names. I think it has been reseted to default. So I'm stuck...What can be done. ?

**Comment by ALTanir** [comment]
calegario commented May 11, 2019- This is very helpful. Thanks for the idea ! ...\Roaming\Mozilla\Firefox\Profiles will find a file called "containers.json". It's important because it contains the containers names and ID's...~ And then can use it to back up by copy-paste

**Comment by FreeWF** [comment]
ALTanir, I've just written before that this file on my computer does not contain the names and IDs of my custom containers. It probably have been reseted to default.

**Comment by FreeWF** [comment]
Why is my "storage.js.migrated" unreadable (probably compressed) when for some people they are readable ? Is there a way to get the URLs ? And for containers.json, is there a way to get a backup of it ? (this because in my case the file was reset to default). Thanks in advance for help...

**Comment by ALTanir** [comment]
FreeWF> ALTanir, ... my computer does not contain the names and IDs of my custom containers. It probably have been reseted to default.
Answer: The roaming'folders contained file "containers.json" views the current activity- which features will shows you when the firefox page running last time & options performable. Then close the firefox and copy-backeup, or read by text. But when if you have lost the elements after app was running- then its .json'files shows you the next feature. And its could not to recover the previous !!
Yes, your says "probably...reseted" means if the reset-progress give you deletion of previous data; no way to find that.
My a story is, Once last year- the multi account container settings was having 35 of list. Per pieces renamed was 1, 2, 3,....35. Then I tried it in a newer version of firefox-upgraded. But unfortunately it converted to default as 4 pieces instead of 35 !! Then all list needs to remake so that determined-self not to use that specific version. And undo the upgrades- but no fund, recovers to previous list. The previous & current version also give me the 4 pieces. I found me as lost of the a big work histories-log in. Then in same day, I was remaking that list of 1 to 35. Then tried up to log in first a container of list. Alas! its fine- magically the previous log in have been continued. All 35 pieces works properly as previous ! So fine- &thats all of story.
Your recovering policies will be same rules -93% confirmed. But for the tragedy of certificate expire, add-ons disables, needs to resign&testing from May3.2019;- may cause the different problems.
Wish we, be committed to request their(add-ons authors) to test&resigned program for every add-ons for to use in specific version of firefox. I have found a resigned .xpi file from favorite and popular author by email(github signup required) after that date immediately. So when you lost the compatibility of extensions, always try it to request by mail, inform, comment etc for resolve to authors.
Thank you
